### DISCUSSION TITLE
problem with `saint_venant_kirchhoff` model

### DISCUSSION THREAD

**Discussion by Sad-Abd** [adtzlr/matadi#136] [General]
Hi @adtzlr,
I was working with `saint_venant_kirchhoff` model and figured out it does not yield expected answers. Then I tried the `Lab` module to illustrate material performance, but I encountered errors, only for `saint_venant_kirchhoff` model, while other models was working smoothly.

Here is my code:
```python
from matadi import Lab, MaterialHyperelastic
from matadi.models import saint_venant_kirchhoff


mat = MaterialHyperelastic(
    saint_venant_kirchhoff, 
    mu = 5.1,
    lmbda = 171.5
)

lab = Lab(mat)
data = lab.run(
    ux=True, 
    stretch_min=1.0, 
    stretch_max=2.0, 
)
fig, ax = lab.plot(data, stability=True)
```

and the error mention that there is a singular matrix that can not get inverted.

**Comment by adtzlr** [comment]
Hi, 

and thanks for reporting. I'll have a look in a few hours.

**Reply by adtzlr** [reply]
The thing is that the SVK model has a limited stretch-range for stability. The only alternative you can do is numeric continuation... (e.g. with `contique`)

For some background infos, here are my slides I presented at the GAMM conference a while ago: 
https://www.researchgate.net/publication/354238304_Work_-_conjugacy_of_stress_and_strain_tensors_and_its_role_in_3D_hyperelastic_isotropic_material_formulation

You can see there that there is no solution for stretches above a certain limit (depends on material parameters and the load case).

Is this research-relevant for your PhD work?

**Reply by Sad-Abd** [reply]
Thanks for your quick response. So is there a range of stretch that the model works? I altered the `stretch_min` and `stretch_max` and didn't get any results. Do you mean that the SVK model can work fine outside `Lab` module and the problem is with drawing its behavior?

Yes, and I'm preparing a paper citing this package. I used Neo-Hookean and Mooney-Rivlin models with no problems and wanted to add SVK too.

**Reply by Sad-Abd** [reply]
I think the problem is with SVK model definition. The strain energy density function has non-zero value at undeformed state:

```python
import numpy as np
from matadi import Lab, MaterialHyperelastic
from matadi.models import saint_venant_kirchhoff

mat_SVK = MaterialHyperelastic(
    saint_venant_kirchhoff, 
    mu = 5.1,
    lmbda = 171.5
)

F = np.eye(3, dtype=float)

print(mat_SVK.function([F])[0])
```

**Reply by adtzlr** [reply]
Sounds great! 

Sorry not to assist more but I have some other paper in the pipeline with figures which would definitely help but I can't share them yet.

About the stretch range: Yes, you're right. And that value is different for the load cases. So it would be better to analyze only one load case in the Lab.

For the non-zero value at the undeformed state: That is definitely a bug - could be related to the conversions from the invariants (Green-Lagrange strain invariants to right Cauchy-Green deformation tensor invariants).

**Reply by Sad-Abd** [reply]
#137 will solve the problem entirely, and the `Lab` also works fine :+1:

**Reply by adtzlr** [reply]
Thanks for figuring that out. Before merging #137 I'll check whats wrong with `dot()`.

**Reply by adtzlr** [reply]
New release 0.1.21 is on PyPI, thanks for your contribution!

**Reply by Sad-Abd** [reply]
I just added my two cents, and it was more like pocket change! :smile:
